autoextract/aio/client.py

```diff
@@ -34,11 +34,101 @@ class Result(List[Dict]):
+        self._query_errors: List[Dict] = list()
+        self._retriable_query_exceptions: List[Dict] = list()
+
+    def _enqueue_error(self, query_result, query_exception):
+        """Enqueue Query-level error.
+
+        Enqueued errors could be:
+            - used in combination with successes with `get_latest_results`
+            - retried using `pending_requests`
+        """
+        self._query_errors.append(query_result)
+        self._retriable_query_exceptions.append(query_exception)
+
+        user_query = query_result["query"]["userQuery"]
+        self.pending_queries.append(user_query)
+
+    def get_latest_results(self):
+        """Get latest results (errors + successes).
+
+        This method could be used to retrieve results
+        when an exception is raised while processing results.
+        """
+        return self._query_successes + self._query_errors
+
+    def process_results(self, query_results):
+        """Process query results.
+
+        Return successful queries and also failed ones.
+
+        If `self._handle_exceptions` is True,
+        this method might raise a `QueryError` exception.
+
+        If multiple `QueryError` exceptions are parsed,
+        the one with the longest timeout is raised.
+
+        Successful requests are saved in `self._query_successes`
+        and they are kept between executions
+        while failures are saved in `self._query_errors`.
+
+        Queries saved in `self._query_errors` are moved to
+        `self.pending_queries` between executions.
+        You can use the first or the n-th result:
+
+            - You can get all queries successfully responded in the first try.
+            - You can get all queries successfully in the n-th try.
+            - You may stop with a partial number of successful queries.
+        """
+        self._reset()
+        for query_result in query_results:
+            if self._handle_query_errors and "error" in query_result:
+                query_exception = QueryError.from_query_result(query_result)
+                if query_exception.retriable:
+                    self._enqueue_error(query_result, query_exception)
+                    continue
+
+            self._query_successes.append(query_result)
+
+        if self._retriable_query_exceptions:
+            # Prioritize exceptions that have retry seconds defined
+            # and get the one with the longest timeout value
+            exception_with_longest_timeout = max(
+                self._retriable_query_exceptions,
+                key=lambda exc: exc.retry_seconds
+            )
+            raise exception_with_longest_timeout
+
+        return self.get_latest_results()
+
+
 async def request_raw(query: Query,
                       api_key: Optional[str] = None,
                       endpoint: str = API_ENDPOINT,
                       *,
                       handle_retries: bool = True,
+                      skip_query_errors: bool = False,
```

Comment:
I'd expect that if skip_query_errors=True is passed, the returned result doesn't contain query errors, not that these errors are no longer retried. 

Unlike errors which are being retried at the moment, query errors may not have a single best strategy for retrial: how much to retry proxy errors, etc. depends on 

* behavior of remote websites
* willingness of user to keep retrying those, as there are diminishing returns

What do you think about having a different argument, something like `query_error_max_retries`? 0 disables retries of query-level errors, other values allow to express willingness of the user to get all results, as compared to costs.
Alternatively, instead of an integer, it can be a tenacity stop condition object, though I don't have a strong opinion on that; int is probably fine.

Reply:
If we go with `query_error_max_retries` (or something like that), CLI may need just this option. I'm not sure there is a need to provide an option to disable retries completely. E.g. 429 messages are a part of the protocol, not errors.

Reply:
I like the max retry count setting, especially if it's an integer because it's easy to implement, easy to understand, and consistent across multiple use cases. It seems like this is the type of argument used by similar API wrapper libraries that need to implement a retry strategy.

I don't think those libraries make any differentiation between paid and free requests, but we should answer this question thinking about our use case: should we differentiate paid from free requests/retries? 

Also, should different Query-level errors have a different max retry count? Should we expand this logic to the Request-level errors as well? Using the same argument or exclusive ones?

If you think this is the path we should follow, I could work on it :) 
Let's discuss about it.

Reply:
Let me provide my thoughts on all that with a little bit of perspective. This is what theoretically I think the library should support:
* Provide meaningfully query-level errors. This means several things:
  - Have a catalogue of all the possible errors that can be received with a clear explanation about what they mean, when they happen, and which is the recommended action on them (e.g. should they be retried, and how should the be retried). Ideally, I would enumerate these errors so that we can reference them. 
  - This catalogue should be complete enough.
  -  Have groups of these errors. Some example groups I'm thinking about: `retriables`, `retriables-with-suggested-time`, `hard-errors`. There can be more, for example, if we need different retry policies by them. 
* Fine-grained support of customization of the retrial policy in the library
  - Being able to change the policy for a particular query-level error
  - Being able to change the policy for a group of errors
  - Being able to configure the behaviour of the retrials. IMO, it would be enough with three parameters for most of the errors: `max-retrials`, `time-between-retrials`, `max-total-time`.
* Command line should have sensible parameters by default, and probably it won't be required so fine-grained configuration support. Probably enough allowing configuring some general parameters.

The former is the ideal state I think the library should get. But I know we are far from that (especially starting with the catalogue of errors, for which we don't have enough knowledge now and is likely to change in the future). And is out of the scope of this PR to reach to that level.

IMO all steps in this PR that get us closer to this desired state are good to me. `query_error_max_retries` is reasonable to me.

Reply:
@victor-torres 

> I don't think those libraries make any differentiation between paid and free requests, but we should answer this question thinking about our use case: should we differentiate paid from free requests/retries?

After looking at exceptions, I think that can be fine not to differentiate them. I think we could differentiate errors which are caused by remote websites or proxies, vs errors which are caused by the API server. All retriable request-level and network errors are caused by the API server, and all retriable query-level errors are caused by remote websites or proxies. So I think that's fine to have logic separated as in this PR: have default, non-customizable behavior for request-level and network errors (as now), and add retries for query-level errors, with a max_retry option.

> Also, should different Query-level errors have a different max retry count? Should we expand this logic to the Request-level errors as well? Using the same argument or exclusive ones?

I think that's fine to leave request-level errors as-is. We should know the best values, because we control the behavior of the API server, and request-level errors and network errors are originated from the API server. 

Retriable query-level errors come from website and from proxies, and so best values can vary, based on websites and on user needs, so it'd be good to allow customization. I think a single max retry value is enough for now, and it can be made more complex later, if we find a need for it.

> I'd go with with a retry count for simplicity, but leave it open for something more "specialized", like tenacity conditions to allow some behavior as described by @ivanprado

+1. We need to do nothing to leave it open for something more specialized, as we can always add a new argument, or reuse the existing one, allowing to receive some tenacity objects.

Reply:
@kmike just to be sure before getting back to code:

- should I implement max retry count and **keep** current tenacity logic as well?
- if the answer to the question above is positive, should I stop to whatever happens first (max number of retries or tenacity stop time), or is there a priority when using those configurations?
- what should be the default number of retries?

Reply:
> should I implement max retry count and keep current tenacity logic as well?

Let's review the current logic.

1. For request-level errors and network errors I think max retry count shouldn't have any effect. I.e. retries which were in a library before this PR should stay the same; the shouldn't depend on max retry parameter.
2. For domain occupied errors there is currently wait time identified in a message + stop_never. I think it should be changed to "stop after max retry count". Ideally, wait time should be picked up from the error message every time, but I'm not sure this is a 100% requirement. If that's hard to implement, using the first value can be ok.
3. For query level errors there is an increasing retry time with jitter + max retry time limit. I'm not sure max retry time limit makes sense if we have max retry time; we just need to pick a good strategy for increasing time between retries. I think it shouldn't be purely exponential, there should be a cap on max time between retrials.

> if the answer to the question above is positive, should I stop to whatever happens first (max number of retries or tenacity stop time), or is there a priority when using those configurations?

I think max number of retrials could be implemented as a part of tenacity stop condition. Regarding logic - see the proposal above.

> what should be the default number of retries?

My vote is for 0.